pkg/rbd/rbd.go

```diff
@@ -0,0 +1,88 @@
+/*
+Copyright 2024 The Rook Authors. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package rbd
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/rook/kubectl-rook-ceph/pkg/exec"
+	"github.com/rook/kubectl-rook-ceph/pkg/k8sutil"
+	"github.com/rook/kubectl-rook-ceph/pkg/logging"
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// List retrieves and displays the Ceph block pools and their associated images.
+func ListRbdVolumes(ctx context.Context, clientsets *k8sutil.Clientsets, operatorNamespace string, clusterNamespace string) {
+	// List Ceph Block Pools
+	blockPoolList, err := clientsets.Rook.CephV1().CephBlockPools(clusterNamespace).List(ctx, v1.ListOptions{})
+	if err != nil {
+		logging.Fatal(fmt.Errorf("failed to list CephBlockPools: %w", err))
+	}
+
+	// Initialize map to store block pool names and associated Rados namespaces
+	blockPoolNames := make(map[string][]string)
+	for _, blockPool := range blockPoolList.Items {
+		blockPoolNames[blockPool.Name] = []string{"--", "--"}
+	}
+
+	// List Ceph Block Pool Rados Namespaces
+	blockPoolNamespaceList, err := clientsets.Rook.CephV1().CephBlockPoolRadosNamespaces(clusterNamespace).List(ctx, v1.ListOptions{})
+	if err != nil {
+		logging.Fatal(fmt.Errorf("failed to list CephBlockPoolRadosNamespaces: %w", err))
+	}
+	for _, blockPoolNameSpace := range blockPoolNamespaceList.Items {
+		var name string
+		// Check if Spec.BlockPoolName exists; otherwise, use ObjectMeta.Name
+		if blockPoolNameSpace.Spec.BlockPoolName != "" {
+			name = blockPoolNameSpace.Spec.Name
+		} else {
+			name = blockPoolNameSpace.ObjectMeta.Name
+		}
+		if _, exists := blockPoolNames[blockPoolNameSpace.Spec.BlockPoolName]; exists {
+			blockPoolNames[blockPoolNameSpace.Spec.BlockPoolName][1] = name
+		}
+	}
+
+	// Retrieve list of RBD images for each pool
+	cmd := "rbd"
+	for poolName := range blockPoolNames {
+		args := []string{"ls", "--pool=" + poolName}
```

Comment:
i dont see we are passing `--namespace` here? Did we verify this by creating multiple namespaces and see if its working as expected?

Reply:
this the relevant scenario? 
```
$ kubectl get CephBlockPoolRadosNamespace -A
NAMESPACE   NAME          PHASE   BLOCKPOOL     AGE
rook-ceph   namespace-a   Ready   replicapool   25m
rook-ceph   namespace-b   Ready   replicapool   78s

```

Reply:
I dont see any output
```
bash-5.1$ rbd ls --pool=replicapool --namespace=namespace-b
bash-5.1$ rbd ls --pool=replicapool --namespace=namespace-a
bash-5.1$ 
```

Reply:
```
bash-5.1$ ceph osd pool ls detail
pool 1 '.mgr' replicated size 1 min_size 1 crush_rule 1 object_hash rjenkins pg_num 32 pgp_num 32 autoscale_mode on last_change 22 lfor 0/0/20 flags hashpspool stripe_width 0 application mgr read_balance_score 1.00
pool 2 'replicapool' replicated size 1 min_size 1 crush_rule 2 object_hash rjenkins pg_num 32 pgp_num 32 autoscale_mode on last_change 22 lfor 0/0/20 flags hashpspool,selfmanaged_snaps stripe_width 0 application rbd read_balance_score 1.00

bash-5.1$ ceph osd pool ls detail
pool 1 '.mgr' replicated size 1 min_size 1 crush_rule 1 object_hash rjenkins pg_num 32 pgp_num 32 autoscale_mode on last_change 22 lfor 0/0/20 flags hashpspool stripe_width 0 application mgr read_balance_score 1.00
pool 2 'replicapool' replicated size 1 min_size 1 crush_rule 2 object_hash rjenkins pg_num 32 pgp_num 32 autoscale_mode on last_change 22 lfor 0/0/20 flags hashpspool,selfmanaged_snaps stripe_width 0 application rbd read_balance_score 1.00

```
I don't see namespace-a/b how to configure it?

Reply:
In my code, I get the `namespace` via kubectl command and not ceph cli https://github.com/rook/kubectl-rook-ceph/pull/333/files#diff-b4c239a28ac425efccdf14790165fceaf4be580365efb6ccc72068a6800958f8R47

Reply:
https://rook.io/docs/rook/latest/CRDs/Block-Storage/ceph-block-pool-rados-namespace-crd/#creating-a-storage-class use this to create images and pvc or you can use rbd cli to create images in the namespace. if we dont pass the --namespace to the RBD CLI we wont list images in the namespace

Reply:
```
$ cat <<EOF | kubectl apply -f -
apiVersion: ceph.rook.io/v1
kind: CephBlockPoolRadosNamespace
metadata:
  name: namespace-a
  namespace: rook-ceph
spec:
  blockPoolName: replicapool
EOF
cephblockpoolradosnamespace.ceph.rook.io/namespace-a created


$  kubectl -n rook-ceph  get cephblockpoolradosnamespace/namespace-a -o jsonpath='{.status.info.clusterID}'
80fc4f4bacc064be641633e6ed25ba7e

 $ cat <<EOF | kubectl apply -f -
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: rook-ceph-block-rados-ns
provisioner: rook-ceph.rbd.csi.ceph.com # csi-provisioner-name
parameters:
  clusterID: 80fc4f4bacc064be641633e6ed25ba7e
  pool: replicapool
EOF
storageclass.storage.k8s.io/rook-ceph-block-rados-ns created


$ cat <<EOF | kubectl apply -f -
apiVersion: ceph.rook.io/v1
kind: CephBlockPool
metadata:
  name: replicapool
  namespace: rook-ceph
spec:
  replicated:
    size: 3
  mirroring:
    enabled: true
    mode: image
    # schedule(s) of snapshot
    snapshotSchedules:
      - interval: 24h # daily snapshots
        startTime: 14:00:00-05:00
EOF
cephblockpool.ceph.rook.io/replicapool created



$ cat <<EOF | kubectl apply -f -
apiVersion: ceph.rook.io/v1
kind: CephBlockPoolRadosNamespace
metadata:
  name: namespace-a
  namespace: rook-ceph # namespace:cluster
spec:
  # The name of the CephBlockPool CR where the namespace is created.
  blockPoolName: replicapool
  mirroring:
    mode: image
    remoteNamespace: namespace-a # default is the same as the local rados namespace
    # schedule(s) of snapshot
    snapshotSchedules:
      - interval: 24h # daily snapshots
        startTime: 14:00:00-05:00
EOF
cephblockpoolradosnamespace.ceph.rook.io/namespace-a configured
```

Working with this procedure and CephBlockPoolRadosNamespace stuck in Progressing Phase and "rbd ls --pool=replicapool --namespace=namespace-a" cli stuck
```
$ kubectl describe CephBlockPoolRadosNamespace namespace-a -n rook-ceph
Name:         namespace-a
Namespace:    rook-ceph
Labels:       <none>
Annotations:  <none>
API Version:  ceph.rook.io/v1
Kind:         CephBlockPoolRadosNamespace
Metadata:
  Creation Timestamp:  2024-11-17T18:06:01Z
  Finalizers:
    cephblockpoolradosnamespace.ceph.rook.io
  Generation:        2
  Resource Version:  2268
  UID:               8078a253-2a23-4787-8b1e-097ba2bdf7c0
Spec:
  Block Pool Name:  replicapool
  Mirroring:
    Mode:              image
    Remote Namespace:  namespace-a
    Snapshot Schedules:
      Interval:    24h
      Start Time:  14:00:00-05:00
Status:
  Info:
    Cluster ID:  80fc4f4bacc064be641633e6ed25ba7e
  Phase:         Progressing
Events:          <none>
```
```
$ kubectl -n rook-ceph exec -it deploy/rook-ceph-tools -- bash
bash-5.1$ rbd ls --pool=replicapool --namespace=namespace-a
```

Reply:
it is not working for me ```
cat <<EOF | kubectl apply -f -
apiVersion: ceph.rook.io/v1
kind: CephBlockPoolRadosNamespace
metadata:
  name: namespace-a
  namespace: rook-ceph
spec:
  blockPoolName: replicapool
EOF

$ kubectl -n rook-ceph  get cephblockpoolradosnamespace/namespace-a -o jsonpath='{.status.info.clusterID}'
80fc4f4bacc064be641633e6ed25ba7e

cat <<EOF | kubectl apply -f -
apiVersion: ceph.rook.io/v1
kind: CephBlockPool
metadata:
  name: replicapool
  namespace: rook-ceph # namespace:cluster
spec:
  failureDomain: osd
  replicated:
    size: 1
    # Disallow setting pool with replica 1, this could lead to data loss without recovery.
    # Make sure you're *ABSOLUTELY CERTAIN* that is what you want
    requireSafeReplicaSize: false
EOF


cat <<EOF | kubectl apply -f -
apiVersion: storage.k8s.io/v1
kind: StorageClass
metadata:
  name: rook-ceph-block-rados-ns
provisioner: rook-ceph.rbd.csi.ceph.com # csi-provisioner-name
parameters:
  clusterID: 80fc4f4bacc064be641633e6ed25ba7e
  pool: replicapool
EOF


cat <<EOF | kubectl apply -f -
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: rbd-pvc
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi
  storageClassName: rook-ceph-block-rados-ns
EOF


$ kubectl get pvc
NAME      STATUS    VOLUME   CAPACITY   ACCESS MODES   STORAGECLASS               VOLUMEATTRIBUTESCLASS   AGE
rbd-pvc   Pending                                      rook-ceph-block-rados-ns   <unset>                 62s

  Type     Reason                Age                From                                                                                                      Message
  ----     ------                ----               ----                                                                                                      -------
  Normal   Provisioning          18s (x7 over 49s)  rook-ceph.rbd.csi.ceph.com_csi-rbdplugin-provisioner-7c8b45d6-ghpk8_8c3f9dcf-a8d5-4cf4-b8fa-d7f00fd6e3d3  External provisioner is provisioning volume for claim "default/rbd-pvc"
  Warning  ProvisioningFailed    18s (x7 over 49s)  rook-ceph.rbd.csi.ceph.com_csi-rbdplugin-provisioner-7c8b45d6-ghpk8_8c3f9dcf-a8d5-4cf4-b8fa-d7f00fd6e3d3  failed to provision volume with StorageClass "rook-ceph-block-rados-ns": rpc error: code = InvalidArgument desc = provided secret is empty
  Normal   ExternalProvisioning  6s (x5 over 49s)   persistentvolume-controller                                                                               Waiting for a volume to be created either by the external provisioner 'rook-ceph.rbd.csi.ceph.com' or manually by the system administrator. If volume creation is delayed, please verify that the provisioner is running and correctly registered.
```